src/services/indexer.ts

```diff
@@ -19,473 +19,467 @@
-      .for(sortedFiles)
-      .process(async programData => {
-        const body = programData.files.map(file => ({ delete: { _id: file.objectId } }));
-        const index = await getNextIndex(programData.program, {
-          isPublic: true,
-          clone: true,
-        });
-
-        try {
-          await client.bulk({
-            index,
-            body,
-          });
-        } catch (e) {
-          logger.error(`Failed bulk delete request: ${JSON.stringify(e)}`, e);
-          throw e;
-        }
-      });
-  }
-
-  async function removeFilesFromRestricted(files: File[]): Promise<void> {
-    const sortedFiles = sortFilesIntoPrograms(files);
-
-    // TODO: Configure ES request concurrency
-    await PromisePool.withConcurrency(20)
-      .for(sortedFiles)
-      .process(async programData => {
-        const body = programData.files.map(file => ({ delete: { _id: file.objectId } }));
-        const index = await getNextIndex(programData.program, {
-          isPublic: false,
-          clone: true,
-        });
-
-        try {
-          await client.bulk({
-            index,
-            body,
-          });
-        } catch (e) {
-          logger.error(`Failed bulk delete request: ${JSON.stringify(e)}`, e);
-          throw e;
-        }
-      });
-  }
-
-  async function deleteIndices(indices: string[]): Promise<void> {
-    if (!indices || !indices.length) {
-      return;
-    }
-    await client.indices.delete({ index: indices });
-
-    // remove the nextIndices and currentIndices references to these indexNames
-    indices.forEach(indexName => {
-      for (const programId in currentIndices.public) {
-        if (currentIndices.public[programId].indexName === indexName) {
-          delete currentIndices.public[programId];
-        }
-      }
-      for (const programId in currentIndices.restricted) {
-        if (currentIndices.restricted[programId].indexName === indexName) {
-          delete currentIndices.restricted[programId];
-        }
-      }
-      for (const programId in nextIndices.public) {
-        if (nextIndices.public[programId].indexName === indexName) {
-          delete nextIndices.public[programId];
-        }
-      }
-      for (const programId in nextIndices.restricted) {
-        if (nextIndices.restricted[programId].indexName === indexName) {
-          delete nextIndices.restricted[programId];
-        }
-      }
-    });
-  }
-
-  return {
-    // By FileDocument
-    indexRestrictedFileDocs,
-    removeRestrictedFileDocs,
-
-    updateRestrictedFile,
-
-    // Public Index Management
-    createEmptyPublicIndices,
-    createEmptyRestrictedIndices,
-    indexPublicFileDocs,
-    deleteIndices,
-    removeFilesFromPublic,
-    removeFilesFromRestricted,
-    release,
-  };
+export const getIndexer = async () => {
+	const rollcall = await getRollcall();
+	const client = await getClient();
+
+	async function indexFiles(index: string, files: FileCentricDocument[]): Promise<void> {
```

Comment:
This function handles sending the indexing request to ES.

It will ensure that the array of files being written to ES index at any time is chunked to a max lenght of 5000 documents. We have had no issues indexing up to 15k documents at a single time, errors have only started now seeing a 30k+ length array. This 5000 document length is therefore a conservative bound.

Reply:
unclear why this function must be defined within the scope of `getIndexer`. should it be abstracted?

Reply:
Needs to be in same scope as the ES client variable `client`

Reply:
I see it now. I'd abstract the function and pass the variable as an argument, lest you want to instantiate a new `indexFiles` function with every execution of `getIndexer` (which smells like a memory leak waiting to happen).